tslearn/neighbors.py

```diff
@@ -243,12 +271,31 @@ def predict(self, X):
         X : array-like, shape (n_ts, sz, d)
             Test samples.
         """
-        check_is_fitted(self, '_X_fit')
-        X = check_array(X, allow_nd=True)
-        X = to_time_series_dataset(X)
-        X_ = to_sklearn_dataset(X)
-        X_ = check_dims(X_, self._X_fit, extend=False)
-        return super(KNeighborsTimeSeriesClassifier, self).predict(X_)
+        if self.metric in self.variable_length_metrics:
+            self._ts_metric = self.metric
+            self.metric = "precomputed"
+
+            if self.metric_params is None:
+                metric_params = {}
+            else:
+                metric_params = self.metric_params
+            check_is_fitted(self, '_ts_fit')
+            X = check_array(X, allow_nd=True, force_all_finite=False)
+            X = to_time_series_dataset(X)
+            if self._ts_metric == "dtw":
+                X_ = cdist_dtw(X, self._ts_fit, **metric_params)
+            elif self._ts_metric == "softdtw":
+                X_ = cdist_soft_dtw(X, self._ts_fit, **metric_params)
+            pred = super(KNeighborsTimeSeriesClassifier, self).predict(X_)
+            self.metric = self._ts_metric
+            return pred
+        else:
+            check_is_fitted(self, '_X_fit')
+            X = check_array(X, allow_nd=True)
+            X = to_time_series_dataset(X)
+            X_ = to_sklearn_dataset(X)
+            X_ = check_dims(X_, self._X_fit, extend=False)
+            return super(KNeighborsTimeSeriesClassifier, self).predict(X_)
 
     def predict_proba(self, X):
```

Comment:
It looks like the predict and predict_proba functions share a lot of common code. Perhaps it would be nice to create a private function and let predict & predict_proba re-use its code?

Reply:
Good idea! Will do.